turbopack/crates/turbo-tasks-fs/src/watcher.rs

```diff
@@ -287,23 +287,39 @@ mod non_recursive_helpers {
         dir_path: &Path,
         root_path: &Path,
     ) -> Result<()> {
-        let mut cur_path = dir_path;
-        loop {
-            start_watching_dir(&mut state.notify_watcher, cur_path, root_path)?;
-
-            let Some(parent_path) = cur_path.parent() else {
-                // this should never happen as we break before we reach the root path
-                anyhow::bail!(
-                    "failed to compute parent path of {cur_path:?} while watching {dir_path:?} in \
-                     root {root_path:?}"
-                );
-            };
-
-            if parent_path == root_path || !state.watched.insert(parent_path.to_path_buf()) {
-                break;
-            }
+        let mut found_watched_ancestor = false;
+
+        // NOTE: `Path::ancestors` yields ancestors from longest to shortest path.
+        let dir_and_ancestor_paths: Vec<_> = [dir_path]
+            .into_iter()
+            .chain(
+                dir_path
+                    .ancestors()
+                    // skip: `ancestors` includes `dir_path` itself, as well as the ancestors, but
+                    // we only want to apply the `take_while` check to parents
+                    .skip(1)
+                    .take_while(|p| {
+                        found_watched_ancestor = *p == root_path || state.watched.contains(*p);
```

Comment:
We can't use `insert` here, and have to do `contains` followed by an `insert` later because `start_watching_dir` could fail. It we don't do this, we could end up with an entry in `watched` that we haven't successfully watched.